### PR TITLE
feat: implement pluginconfig clients (#638)

### PR DESCRIPTION
**Pull request by neverCase**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

- Related issues
#638 
___
### Bugfix
- Description

- How to fix?

___
### New feature or improvement
- Describe the details and related test reports.

___
### Backport patches
- Why need to backport?

- Source branch

- Related commits and pull requests

- Target branch

### CODE REVIEW

**Comment by neverCase** [comment]
@tao12345666333 
hi, i'm ok.
I had add the missing license header.

File: pkg/apisix/plugin_test.go [hunk 1]
```diff
@@ -23,6 +23,7 @@ import (
 	"testing"
 
 	"github.com/apache/apisix-ingress-controller/pkg/metrics"
+
```
Comment (tokers) [inline pkg/apisix/plugin_test.go:26]: Can we sort these imports?

File: pkg/apisix/pluginconfig_test.go [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"sort"
+	"strconv"
+	"strings"
+	"testing"
+
+	"github.com/apache/apisix-ingress-controller/pkg/metrics"
+	v1 "github.com/apache/apisix-ingress-controller/pkg/types/apisix/v1"
```
Comment (tokers) [inline pkg/apisix/pluginconfig_test.go:30]: Ditto, we should sort these imports.

File: pkg/apisix/plugin_test.go (hunk 1 above)
Comment (neverCase) [reply pkg/apisix/plugin_test.go:26]: My local golangci-lint would print these warnings below:
```sh
golangci-lint run
pkg/apisix/plugin_test.go:25: File is not `goimports`-ed with -local github.com/apache/apisix-ingress-controller (goimports)
	"github.com/apache/apisix-ingress-controller/pkg/metrics"
pkg/apisix/schema_test.go:24: File is not `goimports`-ed with -local github.com/apache/apisix-ingress-controller (goimports)
	"github.com/apache/apisix-ingress-controller/pkg/metrics"
```
Finally, I found that if I split the imports then there no warnings left.
Maybe I should ignore it in the next commits.

Comment (tokers) [reply pkg/apisix/plugin_test.go:26]: I think it's caused by the lexicographical orders.

File: pkg/apisix/cache/cache.go [hunk 3]
```diff
@@ -82,4 +88,6 @@ type Cache interface {
 	DeleteConsumer(*v1.Consumer) error
 	// DeleteSchema deletes the specified schema in cache.
 	DeleteSchema(*v1.Schema) error
+	// DeletePluginConfig deletes the specified plugin_config in cache.
+	DeletePluginConfig(config *v1.PluginConfig) error
```
Comment (tao12345666333) [inline pkg/apisix/cache/cache.go:92]:
```suggestion
	DeletePluginConfig(*v1.PluginConfig) error
```

File: pkg/apisix/cache/memdb_test.go [hunk 4]
```diff
@@ -354,7 +354,7 @@ func TestMemDBCacheSchema(t *testing.T) {
 		Name:    "plugins/p1",
 		Content: "plugin schema",
 	}
-	assert.Nil(t, c.InsertSchema(s1), "inserting schema s1")
+	assert.Nil(t, c.InsertSchema(s1), "inserting schema pc1")
```
Comment (tao12345666333) [inline pkg/apisix/cache/memdb_test.go:357]: Why amend it here?

File: pkg/apisix/cache/memdb_test.go [hunk 5]
```diff
@@ -389,3 +389,58 @@ func TestMemDBCacheSchema(t *testing.T) {
...
+	c, err := NewMemDBCache()
+	assert.Nil(t, err, "NewMemDBCache")
+
+	pc1 := &v1.PluginConfig{
+		Metadata: v1.Metadata{
+			ID:   "1",
+			Name: "name1",
+		},
+	}
+	assert.Nil(t, c.InsertPluginConfig(pc1), "inserting plugin_config s1")
```
Comment (tao12345666333) [inline pkg/apisix/cache/memdb_test.go:403]: There is no s1.

File: pkg/apisix/cache/memdb_test.go [hunk 6]
```diff
@@ -389,3 +389,58 @@ func TestMemDBCacheSchema(t *testing.T) {
...
+			Name: "name2",
+		},
+	}
+	s3 := &v1.PluginConfig{
+		Metadata: v1.Metadata{
+			ID:   "3",
+			Name: "name3",
+		},
+	}
+	assert.Nil(t, c.InsertPluginConfig(pc2), "inserting plugin_config s2")
```
Comment (tao12345666333) [inline pkg/apisix/cache/memdb_test.go:421]: And no s2

File: pkg/apisix/pluginconfig.go [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+	)
+	if err := pc.cluster.HasSynced(ctx); err != nil {
+		return nil, err
+	}
+	body, err := json.Marshal(obj)
+	if err != nil {
+		return nil, err
+	}
+	url := pc.url + "/" + obj.ID
+	log.Debugw("updating username", zap.ByteString("body", body), zap.String("url", url))
```
Comment (tao12345666333) [inline pkg/apisix/pluginconfig.go:220]: typo. no "username"

**Review by tao12345666333 — CHANGES_REQUESTED** [review]

**Comment by tao12345666333** [comment]
@neverCase  Sorry for delay. I leave some comments, I hope you can make changes

**Comment by neverCase** [comment]
> @neverCase Sorry for delay. I leave some comments, I hope you can make changes

ok, i will change these and update right now

**Comment by neverCase** [comment]
@tao12345666333 
hi, i'm ready

**Review by tao12345666333 — APPROVED** [review]
LGTM, Thanks!

**Comment by tao12345666333** [comment]
ping @gxthrj @tokers  for review, thanks!

**Review by tokers — COMMENTED** [review]
Mostly LGTM but some import block style problems.

**Comment by tao12345666333** [comment]
> Mostly LGTM but some import block style problems.

@neverCase can you improve the code style?

**Comment by neverCase** [comment]
> > Mostly LGTM but some import block style problems.
> 
> @neverCase can you improve the code style?

I had revert it

**Comment by tao12345666333** [comment]
#788  has fixed the e2e error. It's green!

**Review by gxthrj — APPROVED** [review]

**Comment by tao12345666333** [comment]
Good job!

**Pull request merged**